Start some docs for the operator

tombentley opened the pull request:
### Type of change

- Documentation

### Description

This PR provides the basic structure of some end user documentation for the operator, as well as providing some more detailed doucmentation for the TLS work in #2070.

Part of #1828 and #1832.

tombentley commented on docs/_assets/trademarks.adoc:
```diff
@@ -1,8 +1,10 @@
 ifdef::include-fortanix-dsm-kms[]
 * Fortanix and Data Security Manager are trademarks of Fortanix, Inc.
 endif::[]
-* Apache Kafka is a registered trademark of The Apache Software Foundation.
```
It's an unordered list, but it still seemed odd that Kafka was last.

tombentley commented:
@k-wall @PaulRMellor PTAL.

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-auth.adoc:
```diff
@@ -0,0 +1,29 @@
+
+The TLS client certificate you provide will need to have been issued by a Certificate Authority (CA) that's trusted by the proxied cluster.
+
+.Example `KafkaService` configured to authenticate using the TLS client certificate in the `Secret` called `tls-cert-for-kafka.example.com`
```
```suggestion
This example configures a `KafkaService` to use a TLS client certificate stored in a `Secret` named `tls-cert-for-kafka.example.com`.

.Example `KafkaService` configuration with TLS client authentication
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-auth.adoc:
```diff
@@ -0,0 +1,29 @@
+      name: tls-cert-for-kafka.example.com # <3>
+    # ...
+----
+<1> The `certificateRef` object identifies the TLS client certificate to use.
```
```suggestion
<1> The `certificateRef` property identifies the TLS client certificate to use.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-auth.adoc:
```diff
@@ -0,0 +1,29 @@
+    # ...
+----
+<1> The `certificateRef` object identifies the TLS client certificate to use.
+<2> The `kind` is optional and defaults to `Secret`. The `Secret` should have `type: kubernetes.io/tls`.
```
any other acceptable kinds?

tombentley replied:
Not right now

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-by-bootstrap.adoc:
```diff
@@ -0,0 +1,23 @@
+[id='con-configuring-kafkaservice-bootstrap-{context}']
+= Configuring the proxied Kafka cluster
+
+.Example `KafkaService` configured for connections to `kafka.example.com` port 9092
```
```suggestion
This example configures a `KafkaService` to connect to `kafka.example.com` on port 9092.

.Example `KafkaService` configuration
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-by-bootstrap.adoc:
```diff
@@ -0,0 +1,23 @@
+      end: 5 # <5>
+  # ...
+----
+<1> The `bootstrapServers` property is a comma-separated list of addresses in `<host>:<port>` syntax. It is recommended to specify multiple brokers if possible, to enable clients to connect even when a broker is not running.
```
```suggestion
<1> The `bootstrapServers` property is a comma-separated list of addresses in `<host>:<port>` format. Including multiple broker addresses helps clients connect when one is unavailable.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-by-bootstrap.adoc:
```diff
@@ -0,0 +1,23 @@
+  # ...
+----
+<1> The `bootstrapServers` property is a comma-separated list of addresses in `<host>:<port>` syntax. It is recommended to specify multiple brokers if possible, to enable clients to connect even when a broker is not running.
+<2> `nodeIdRanges` declares the ids of all the broker nodes in the Kafka cluster
+<3> `name` is optional, but specifying it can make errors easier to diagnose.
+<4> The start of the id range, inclusive.
+<5> The end of the id range, inclusive.
```
```suggestion
<2> `nodeIdRanges` declares the IDs of all the broker nodes in the Kafka cluster
<3> `name` is optional, but specifying it can make errors easier to diagnose.
<4> The start of the ID range, inclusive.
<5> The end of the ID range, inclusive.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,31 @@
+[id='con-configuring-kafkaservice-cipher-{context}']
+= Controlling the TLS ciphers used
```
```suggestion
= Controlling TLS cipher suites
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,31 @@
+[id='con-configuring-kafkaservice-cipher-{context}']
+= Controlling the TLS ciphers used
+
+Some cipher suites, which are enabled by default for interoperability purposes, are no longer consider secure.
```
Perhaps this needs more context?

_Some older cipher suites enabled by default for TLS connections between clients and virtual clusters are now considered insecure._

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,31 @@
+
+Some cipher suites, which are enabled by default for interoperability purposes, are no longer consider secure.
+
+IMPORTANT: It is considered good practice to disable insecure cipher suites. You can only do this is the proxied Kafka cluster you want to connect to supports the cipher suites that you want to use.
```
```suggestion
IMPORTANT: It is good practice to disable insecure cipher suites. Only do this if the proxied Kafka cluster supports the cipher suites you want to allow.
```

PaulRMellor replied:
Should this be a note or in the main body to introduce what we're showing in the config?

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,31 @@
+
+IMPORTANT: It is considered good practice to disable insecure cipher suites. You can only do this is the proxied Kafka cluster you want to connect to supports the cipher suites that you want to use.
+
+The cipher suites that are enabled by default can depend on the JVM used in the proxy image and the TLS protocol version which is negotiated.
```
```suggestion
The cipher suites enabled by default depend on the JVM used in the proxy image and the TLS protocol version that is negotiated.
```

PaulRMellor replied:
I would change the order and bring this together with the sentence on _some cipher suites_

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,31 @@
+<1> The `cipherSuites` object configures the cipher suites.
+<2> `allowed` lists the cipher suites which are permitted.
+
+The `cipherSuites` object also supports `denied` if you prefer to exclude the unpermitted cipher suites, rather than include the permitted cipher suites.
```
```suggestion
The `cipherSuites` property also supports `denied`, if you prefer to list the cipher suites to exclude instead.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+[id='con-configuring-kafkaservice-protocol-{context}']
+= Controlling the TLS version used
```
```suggestion
= Controlling TLS versions
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+[id='con-configuring-kafkaservice-protocol-{context}']
+= Controlling the TLS version used
+
+Older versions of TLS (and SSL before it) are now considered insecure.
```
```suggestion
Some older versions of TLS (and SSL before it) are now considered insecure.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+
+Older versions of TLS (and SSL before it) are now considered insecure.
+
+IMPORTANT: It is considered good practice to disable insecure versions of the protocol. You can only do this is the proxied Kafka cluster you want to connect to supports the version that you want to use.
```
```suggestion
IMPORTANT: It is good practice to disable insecure protocol versions. Only do this if the proxied Kafka cluster supports the version that you want to use.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+        - TLSv1.3
+        - TLSv1.2
+----
+<1> The `protocols` object configures the TLS procotol versions
```
```suggestion
<1> The `protocols` property configures the TLS protocol versions
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+<1> The `protocols` object configures the TLS procotol versions
+<2> `allowed` lists the versions of TLS which are permitted.
+
+The `protocols` object also supports `denied` if you prefer to exclude the unpermitted versions, rather than including the permitted versions.
```
```suggestion
The `protocols` property also supports `denied`, if you prefer to list the versions to exclude instead.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+= Trusting the proxied cluster's TLS certificates
+
+By default, when connecting to the proxied cluster using TLS, the proxy will use the platform's default trust store.
+This will usually work if the proxied cluster's TLS certificates were issued by a well-known public Certificate Authority (CA).
```
```suggestion
This usually works if the proxied cluster's TLS certificates were issued by a well-known public Certificate Authority (CA).
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+      key: trusted-cas.pem # <4>
+    # ...
+----
+<1> The `trustAnchorRef` object references a separate Kubernetes resource which contains the CA certificates to be trusted
```
```suggestion
<1> The `trustAnchorRef` property references a separate Kubernetes resource which contains the CA certificates to be trusted
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+    # ...
+----
+<1> The `trustAnchorRef` object references a separate Kubernetes resource which contains the CA certificates to be trusted
+<2> The `kind` is optional and defaults to `ConfigMap`.
```
shall we say what it's for and when to use it too?

tombentley replied:
We have `kind` because we anticipate that eventually we might will support other kinds of resource. Because of the defaults it can be omitted. But including it explicitly makes it more obvious what's being referred to.

So I think it's better for users if they include it, because their resources are more understandable that way. But there's not much more to say about it right now, because we only support `ConfigMap`.

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+
+IMPORTANT: It is good practice to configure trust explicitly, even when proxied cluster's TLS certificates are signed by a public CA.
+
+.Example `KafkaService` configured to trust TLS certificates signed by any of the CAs in the `trusted-cas.pem` entry of the `ConfigMap` called `trusted-cas`
```
```suggestion
This example configures a `KafkaService` to trust TLS certificates signed by any Certificate Authority (CA) listed in the `trusted-cas.pem` entry of the `ConfigMap` named `trusted-cas`.

.Example `KafkaService` configuration for trusting certificates
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+----
+<1> The `trustAnchorRef` object references a separate Kubernetes resource which contains the CA certificates to be trusted
+<2> The `kind` is optional and defaults to `ConfigMap`.
+<3> The `name` is the name of the resource of the given `kind`. This resource must exist in the same namespace as the `KafkaService`
```
```suggestion
<3> The `name` of the resource of the given `kind`. This resource must exist in the same namespace as the `KafkaService`
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,28 @@
+<1> The `trustAnchorRef` object references a separate Kubernetes resource which contains the CA certificates to be trusted
+<2> The `kind` is optional and defaults to `ConfigMap`.
+<3> The `name` is the name of the resource of the given `kind`. This resource must exist in the same namespace as the `KafkaService`
+<4> The `key` identifies the entry in the given resource. The corresponding value should be a PEM-encoded set of CA certificates.
```
```suggestion
<4> The `key` identifies the entry in the given resource. The corresponding value must be a PEM-encoded set of CA certificates.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-protocol.adoc:
```diff
@@ -0,0 +1,29 @@
+
+IMPORTANT: It is considered good practice to disable insecure versions of the protocol. You can only do this is the proxied Kafka cluster you want to connect to supports the version that you want to use.
+
+.Example `KafkaService` configured so that the proxy will negotiate only a TLS v1.3 connection to `kafka.example.com`.
```
```suggestion
This example configures a `KafkaService` to allow only TLS v1.3 and v1.2 when connecting to `kafka.example.com`.

.Example `KafkaService` with restricted TLS protocol versions
```

tombentley commented on docs/assemblies/assembly-operator-overview.adoc:
```diff
@@ -0,0 +1,17 @@
+= Kroxylicious Operator overview
+
+[role="_abstract"]
+Kroxylicious Proxy is an Apache Kafka protocol-aware ("Layer 7") proxy designed to enhance Kafka-based systems.
```
I think it's worth having both these sentences (i.e. this and the next) in the operator guide, because they're defining the terms used in the rest of the guide.

tombentley commented on docs/modules/con-about-operator-guide.adoc:
```diff
@@ -0,0 +1,6 @@
+= About this guide
+
+This guide covers using the *Kubernetes operator* for Kroxylicious, including configuration, security and operation.
+Other guides should be consulted if you want run the proxy on platforms other than Kubernetes or to develop plugins for Kroxylicious.
```
The reason I added this is because we can't assume that the reader has navigated to this guide from the home page of the site. For example if they've just followed a link from somewhere randomly on the web. So they might not be aware what other documentation exists, and better match what they're trying to achieve.

tombentley commented:
@PaulRMellor thanks!

I've added your suggestions. I also reworked the sections on TLS protocols and cipher suites. Please let me know if you think it's better now.

PaulRMellor commented on docs/assemblies/assembly-operator-configuring-kafkaprotocolfilters.adoc:
```diff
@@ -0,0 +1,61 @@
+
+For both `secret` and `configmap` providers:
+
+* the path is interpreted as the name of a `Secret` or `ConfigMap` resource in the same namespace as the `KafkaProtocolFilter` resource.
+* the key is interpreted as a key in the `data` property of the `Secret` or `ConfigMap` resource.
```
```suggestion
* The path is interpreted as the name of a `Secret` or `ConfigMap` resource in the same namespace as the `KafkaProtocolFilter` resource.
* The key is interpreted as a key in the `data` property of the `Secret` or `ConfigMap` resource.
```

PaulRMellor commented on docs/modules/con-about-operator-guide.adoc:
```diff
@@ -0,0 +1,6 @@
+= About this guide
+
+This guide covers using the *Kroxylicious Operator* to configure, deploy, secure, and operate the Kroxylicious proxy on Kubernetes.
+Other guides should be consulted if you want run the proxy on platforms other than Kubernetes or to develop plugins for Kroxylicious.
```
```suggestion
Refer to other Kroxylicious guides for information on running the proxy outside Kubernetes or for advanced topics such as plugin development.
```

PaulRMellor replied:
Maybe we also need to condition this as plugin development not a downstream thing?

PaulRMellor commented on docs/modules/con-about-proxy-guide.adoc:
```diff
@@ -2,5 +2,6 @@
 
-This guide covers using the proxy, including configuration, security and operation.
+This guide covers installing, configuring, securing, and operating the Kroxylicious Proxy in a non-Kubernetes environment.
+To run the proxy on Kubernetes, see the separate guide for the Kroxylicious Operator.
 Other guides should be consulted if you want to developing plugins for Kroxylicious.
```
```suggestion
Refer to other Kroxylicious guides for information on running the proxy on Kubernetes using the Kroxylicious Operator, or for advanced topics such as plugin development.
```

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-cipher.adoc:
```diff
@@ -0,0 +1,37 @@
+[id='con-configuring-kafkaservice-cipher-{context}']
+= Controlling TLS cipher suites
+
+A cipher suite is a set of cryptographic algorithms that together provide the security guarantees offered by TLS.
```
This intro reads much better now. :+1:

PaulRMellor commented on docs/modules/configuring/con-configuring-kafkaservice-trust.adoc:
```diff
@@ -0,0 +1,30 @@
+[id='con-configuring-kafkaservice-trust-{context}']
+= Trusting the proxied cluster's TLS certificates
+
+By default, when connecting to the proxied cluster using TLS, the proxy will use the platform's default trust store.
+This usually works if the proxied cluster's TLS certificates were issued by a well-known public Certificate Authority (CA).
+It will not work if the proxied cluster's TLS certificates were issued by a private Certificate Authority.
```
```suggestion
By default, the proxy uses the platform's default trust store when connecting to the proxied cluster over TLS.  
This works if the cluster's TLS certificates are signed by a well-known public Certificate Authority (CA), but fails if they’re signed by a private CA instead.
```

PaulRMellor replied:
Don't think we need to be so cautious with _usually_?

PaulRMellor reviewed (approved):
Thanks for the addressing the comments, Tom.
Few more suggestions, but looks v. good to me. 
I've approved, but if you intend to expand more in the same PR, I can re-review again

tombentley commented:
@k-wall still waiting for your review. Thanks!

k-wall commented on docs/assemblies/assembly-operator-api.adoc:
```diff
@@ -0,0 +1,44 @@
+`KafkaProtocolFilter`::
+Specifies filter mechanisms for use with a virtual cluster.
+`Secret`::
+`KafkaService` and `KafkaProtocolFilter` resources may reference a `Secret` to provide security-sensitive data such as TLS certificates or passwords.
```
The work of #2081 will mean KafkaVirtualCluster should be added to this list.

tombentley replied:
I think #2081 will be a whole other PR.

k-wall reviewed (approved):
That's a great start, thank you @tombentley
